src/witness_generator/machines/block_machine.rs

```diff
@@ -0,0 +1,556 @@
+            };
+            *v == expected
+        })
+        .then_some((poly.to_string(), period))
+}
+
+impl Machine for BlockMachine {
+    fn process_plookup(
+        &mut self,
+        fixed_data: &FixedData,
+        fixed_lookup: &mut FixedLookup,
+        kind: IdentityKind,
+        left: &[Result<AffineExpression, EvalError>],
+        right: &SelectedExpressions,
+    ) -> Option<EvalResult> {
+        if is_simple_poly(right.selector.as_ref()?)? != self.selector
+            || kind != IdentityKind::Plookup
+        {
+            return None;
+        }
+        let previous_len = self.rows() as usize;
+        Some({
+            let result = self.process_plookup_internal(fixed_data, fixed_lookup, left, right);
+            self.bit_constraints.clear();
+            result.map_err(|e| {
+                // rollback the changes.
+                for col in self.data.values_mut() {
+                    col.truncate(previous_len)
+                }
+                e
+            })
+        })
+    }
+
+    fn witness_col_values(
+        &mut self,
+        fixed_data: &FixedData,
+    ) -> HashMap<String, Vec<AbstractNumberType>> {
+        std::mem::take(&mut self.data)
+            .into_iter()
+            .map(|(id, values)| {
+                let mut values = values
+                    .into_iter()
+                    .map(|v| v.unwrap_or_default())
+                    .collect::<Vec<_>>();
+
+                values.resize(fixed_data.degree as usize, 0.into());
+                (fixed_data.name(id), values)
+            })
+            .collect()
+    }
+}
+
+impl BlockMachine {
+    /// Extends the data with a new block.
+    fn append_new_block(&mut self) {
+        for col in self.data.values_mut() {
+            col.resize_with(col.len() + self.block_size, || None);
+        }
+    }
+
+    fn rows(&self) -> DegreeType {
+        self.data.values().next().unwrap().len() as DegreeType
+    }
+
+    fn process_plookup_internal(
+        &mut self,
+        fixed_data: &FixedData,
+        fixed_lookup: &mut FixedLookup,
+        left: &[Result<AffineExpression, EvalError>],
+        right: &SelectedExpressions,
+    ) -> EvalResult {
+        // First check if we already store the value.
+        if left
+            .iter()
+            .all(|v| v.as_ref().ok().map(|v| v.is_constant()) == Some(true))
+        {
+            return Ok(vec![]);
+            // TOOD check that they really exist (maybe just check the last row)
+        }
+
+        let old_len = self.rows();
+        if old_len + self.block_size as u64 >= fixed_data.degree {
+            return Err("Rows in block machine exhausted.".to_string().into());
+        }
+        self.append_new_block();
+        let mut outer_assignments = vec![];
+
+        // TODO this assumes we are always using the same lookup for this machine.
+        let sequence = self.processing_sequence_cache.get_processing_sequence(left);
+
+        let mut errors = vec![];
```

Comment:
This could be fixed by the `EvalResult` refactor we discussed previously? Not blocking